ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,407 @@ let order_f (_, vm_rec) =
+
+(** Try to get a spread evenly plan for anti-affinity VMs (for each anti-affinity group, the number of 
+    running VMs from that group are spread evenly in all the rest hosts in the pool):
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, select a host which can run it, and which has the least VMs in the
+       same anti-affinity group running on it, for the hosts with the same number of running VMs in
+       that group, pick the one with the least free memory. *)
+let rec vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+    vms_increasing vm_can_boot_on_host grp_ranked_hosts mapping =
+  debug "vm_anti_affinity_spread_evenly_plan" ;
+  log_vms_increasing ~__context vms_increasing ;
+  log_hosts_increasing ~__context hosts_increasing ;
+  log_grp_ranked_hosts ~__context grp_ranked_hosts ;
+  match vms_increasing with
+  | [] ->
+      mapping
+  | (vm, vm_size) :: remaining_vms -> (
+      let host_selector hosts =
+        hosts_increasing
+        |> List.filter (fun (h, size) -> List.mem h hosts && size >= vm_size)
+        |> List.map fst
+        |> fst_host_can_run vm_can_boot_on_host vm
+      in
+      match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
+      | Some (`AntiAffinity group) -> (
+          debug
+            "Spread evenly plan: try to plan for the anti-affinity VM (%s, %s)"
+            (Ref.string_of vm)
+            (Db.VM.get_name_label ~__context ~self:vm) ;
+          try
+            let h =
+              Xapi_vm_helpers.select_host_from_ranked_groups vm host_selector
+                (VMGroupMap.find group grp_ranked_hosts
+                |> IntMap.to_seq
+                |> List.of_seq
+                |> List.fast_sort (fun (a, _) (b, _) -> compare a b)
+                   (* Make sure the ranked list is sorted in terms of vm_count increasingly *)
+                |> List.map snd
+                )
+            in
+            debug "Spread evenly plan: find h: (%s %s)" (Ref.string_of h)
+              (Db.Host.get_name_label ~__context ~self:h) ;
+            let h_size =
+              Int64.sub
+                (List.assoc h hosts_increasing)
+                (List.assoc vm vms_increasing)
+            in
+            debug
+              "Spread evenly plan: choose the host with the least free memory \
+               which can run the vm: %s"
+              (Db.Host.get_name_label ~__context ~self:h) ;
+
+            let hosts_increasing =
+              Binpack.insert bigger_than (h, h_size)
+                (List.remove_assoc h hosts_increasing)
+            in
+            vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+              remaining_vms vm_can_boot_on_host
+              (grp_ranked_hosts |> update_anti_affinity_grp_ranked_hosts group h)
```

Comment:
If the data structure could be `HostMap`, then the update would be as simple as:
```
HostMap.update h Option.(value ~default:1 (map (fun n -> n + 1))) hosts
```

And the only cost is to generate the ranks (required by `select_host_from_ranked_groups`) by:
```
HostMap.fold (fun h vm_cnt acc ->
  IntMap.update vm_cnt Option.(value ~default:[h] (map (fun hs -> h::hs))) acc
) hosts IntMap.empty
```

Reply:
Thanks, I would prefer the current implementation:
1. Don't need to regenerate the whole ranked_list everytime a VM is planned on a host.
2. Reuse(share) the same code: `Xapi_vm_helpers.rank_hosts_by_vm_cnt_in_group` with VM start, this can make sure they behave the same as code changes.

Reply:
I think updating the `grp_ranked_hosts` is really painful.

Reply:
The time complexity of `update_anti_affinity_grp_ranked_hosts` is:
` log N + (N * log N * log N) + log N * (N * log N  + log N)`

But after choosing a proper data structure, the time complexity is:
update: `log N`
re-generate:  `N * log N`

Reply:
In current implementation, I think the time complexity of `update_anti_affinity_grp_ranked_hosts ` is 5N (the number of hosts in the pool):
`locate_host` is N

the operation to update ranked_grp for the host's original rank:
```
          ranked_hosts
          |> IntMap.update vm_cnt_on_host (fun hosts ->
                 match
                   get_hosts hosts |> List.filter (fun h -> fst h <> host)
                 with
                 | [] ->
                     None
                 | hs ->
                     Some hs
             )
```
IntMap to locate `vm_cnt_on_host `: log N
Filter it out from the original list: N

then add it to the next rank:
```
          |> IntMap.update (vm_cnt_on_host + 1) (fun hosts ->
                 Some
                   (Binpack.insert bigger_than (host, updated_host_size)
                      (get_hosts hosts)
                   )
             )
```
time complexity:
IntMap to locate `vm_cnt_on_host +1 `: log N
Insert it into the ordered list: N

For the other VM groups(at most 4 now)
```
          ranked_hosts
          |> IntMap.update vm_cnt_on_host (fun hosts ->
                 Some
                   (Binpack.insert bigger_than (host, updated_host_size)
                      (List.remove_assoc host (get_hosts hosts))
                   )
             )
```
IntMap to locate `vm_cnt_on_host `: log N
Update it in the ordered list: N

_______________________________

While for re-generating from a HostMap (host -> numOfVMs), I think the time complexity will be more than N * log N, for each rank, we still need to sort the hosts in terms of each host's free memory (we at least need log N to insert a host into a sorted list, but if insert with List, I think it might be N), then the overall time complexity will be: `N * log N * log N` or even `N * log N * N`.